src/main/java/com/hierynomus/smbj/share/FileOutputStream.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (C)2016 - SMBJ Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hierynomus.smbj.share;
+
+import com.hierynomus.mserref.NtStatus;
+import com.hierynomus.mssmb2.SMB2FileId;
+import com.hierynomus.mssmb2.messages.SMB2WriteRequest;
+import com.hierynomus.mssmb2.messages.SMB2WriteResponse;
+import com.hierynomus.protocol.commons.concurrent.Futures;
+import com.hierynomus.smbj.ProgressListener;
+import com.hierynomus.smbj.common.SMBApiException;
+import com.hierynomus.smbj.connection.Connection;
+import com.hierynomus.smbj.session.Session;
+import com.hierynomus.smbj.transport.TransportException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Future;
+
+public class FileOutputStream extends OutputStream {
+
+    private TreeConnect treeConnect;
+    private SMB2FileId fileId;
+    private Session session;
+    private Connection connection;
+    private int maxWriteSize;
+    private ProgressListener progressListener;
+
+    private byte[] buf;
+    private long offset = 0;
+    private int curr = 0;
+    private boolean isClosed = false;
+    private List<Future<SMB2WriteResponse>> responses = new ArrayList<>();
+    private int availableCredits;
+    private static final Logger logger = LoggerFactory.getLogger(FileOutputStream.class);
+
+    public FileOutputStream(SMB2FileId fileId, TreeConnect treeConnect, ProgressListener progressListener) {
+        this.treeConnect = treeConnect;
+        this.fileId = fileId;
+        this.session = treeConnect.getSession();
+        this.connection = session.getConnection();
+        this.progressListener = progressListener;
+        this.maxWriteSize = connection.getNegotiatedProtocol().getMaxWriteSize();
+        this.buf = new byte[maxWriteSize];
+        this.availableCredits = connection.getConnectionInfo().getSequenceWindow().available();
```

Comment:
This won't work, because you the sequence window might actually be depleted by another thread while you're running.